src/org/parosproxy/paros/extension/state/ExtensionState.java

```diff
@@ -67,11 +77,50 @@ public void hook(ExtensionHook extensionHook) {
 	    // ZAP: Changed to check if there is a view.
 	    if (getView() != null) {
 		    extensionHook.getHookMenu().addEditMenuItem(extensionHook.getHookMenu().getMenuSeparator());
+		    extensionHook.getHookMenu().addEditMenuItem(getMenuZAPMode());
 		    extensionHook.getHookMenu().addEditMenuItem(getMenuSessionTrackingEnable());
 		    extensionHook.getHookMenu().addEditMenuItem(getMenuResetSessionState());
 	    }
 	}
 
+	private JMenuItem getMenuZAPMode(){
+		if (menuZAPMode == null) {
+			menuZAPMode = new JMenu(Constant.messages.getString("menu.edit.zapmode"));
+			int modeCount = 0;
+			for (Mode modeType : Mode.values()) {
+			menuItemsMode[modeCount] = addZAPModeMenuItem(modeType);
+  			menuZAPMode.add(menuItemsMode[modeCount]);
+			modeCount++;
+			}
+			modeCount = 0;
+			for(Mode modeType : Mode.values()){
+			addZAPModeEventListeners(modeType,menuItemsMode[modeCount],menuItemsMode);
+			modeCount++;
+			}
+		}
+		return menuZAPMode;
+	}
+
+	private void addZAPModeEventListeners(final Mode modeType, final JCheckBoxMenuItem checkBox, final JCheckBoxMenuItem[] menuItemsMode){
+		 checkBox.addActionListener(new java.awt.event.ActionListener() {
+		 @Override
+		 public void actionPerformed(java.awt.event.ActionEvent evt) {
+	 	 	Control.getSingleton().setMode(modeType);
+	 		for (int modeCount=0;modeCount<Mode.values().length;modeCount++) {
+		 	menuItemsMode[modeCount].setState(false);
+	 		}
+	 		checkBox.setState(true);
+	 		if(View.isInitialised()){
+	 			View.getSingleton().getMainFrame().getMainToolbarPanel().setMode(modeType);
+	 		}
+		 }
+		});
+	}
+
+	private JCheckBoxMenuItem addZAPModeMenuItem(Mode modeType){
+		final JCheckBoxMenuItem modeItem = new JCheckBoxMenuItem(modeType.name().substring(0,1).toUpperCase()+modeType.name().substring(1)+" Mode");

```

Comment:
This should be internationalised, since there are already resource messages (e.g. "view.toolbar.mode.safe.select") it could use those.